FHT: Redirect to start of the flow when user already had a free trial

vykes-mac opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->
Users can go to `/trialAcknowledge` and click the `Start trial` button even though they had already completed a trial.
This PR redirects to start of flow to prevent creating an empty site with trial button
Related to #

## Proposed Changes

* Redirect to start of the flow if user not eligible

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Enroll in a hosting trial or use an account that had already enrolled
* go to `/new-hosted-site/trialAcknowledge`
* Verify you are redirected to start of the flow.


## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

vykes-mac committed:
show inelgibility message for users who already had a trial

vykes-mac commented:
Maybe the message can be improved, any suggestion is appreciated

vykes-mac committed:
change copy

zaguiini commented:
I see what this PR is doing (and I appreciate the effort), but I'm failing to see why we need this. The user won't see this step directly unless they manually add the plan to the cart and fool the onboarding mechanism, and on top of that we have a check on the BE to avoid FHT spoofing... so, this feels overprotective 😓 Unless I'm missing something, of course.

vykes-mac commented:
> I see what this PR is doing (and I appreciate the effort), but I'm failing to see why we need this. The user won't see this step directly unless they manually add the plan to the cart and fool the onboarding mechanism, and on top of that we have a check on the BE to avoid FHT spoofing... so, this feels overprotective 😓 Unless I'm missing something, of course.

I take your point but currently If we show the trial screen and the user clicks start trail it will create a site, is this something we want? I'm looking at it from the perspective of browsing /trialAcknowledge directly by passing the plan stage, most users might not do that though. Does this PR adds maintenance overhead for a little UI friendliness/protection?

zaguiini commented:
I see @vykes-mac, what if we just redirect them to the start of the flow then?

vykes-mac committed:
redirect to start of the flow if user is not eligible

zaguiini commented on client/landing/stepper/declarative-flow/internals/steps-repository/trial-acknowledge/hosting-trial-acknowledge.tsx:
```diff
@@ -49,10 +50,14 @@ const CallToAction = ( { onStartTrialClick }: CallToActionProps ) => {
 	const planFeatures =
 		plan && 'getPlanCompareFeatures' in plan ? plan.getPlanCompareFeatures?.() ?? [] : [];
 
+	if ( ! isEligible ) {
+		return window.location.assign( '/setup/new-hosted-site' );
+	}
```
Can we do that inside an `useLayoutEffect` instance? The reason is that every component should return a valid node, and this functional call doesn't do that.

vykes-mac replied:
yeah this make sense

p-jackson replied:
I've just pushed an update to move the redirection to an effect callback.

vykes-mac replied:
haha I was pushing the same time an was wondering why there was a collision. I was going to put it in the `useLayoutEffect` in the `new-hosted-flow.ts` file

p-jackson replied:
>  I was going to put it in the useLayoutEffect in the new-hosted-flow.ts file

Oh interesting! Yes that actually does look like a good place to put that. Similar to how we redirected logged out users. Like "you should never be here". The `useSideEffect` function even passes the current step name, so it's already all set up for us 👍

zaguiini reviewed (approved):
Looks good 👍

p-jackson committed:
HostingTrialAcknowledgeInternal always returns a valid React node

The side effect of redirecting is moved into a useEffect block, and
while that's resolving the component renders no elements.

p-jackson committed:
Fix type error

p-jackson commented:
@vykes-mac pushed another change, there was a missing type import. Hopefully good to go now, I'm planning to deploy it.

p-jackson commented:
@vykes-mac scratch the above, I like your idea of putting it in `new-hosted-flow.ts`!

vykes-mac committed:
move eligibility check to useLayoutEffect of the hosting flow

p-jackson reviewed (approved):
lgtm
